Adding multi segment route parameters

stephengaito opened the pull request:
Daniel,

I have a need to build a simple webserver based on your RESTful spin project.

However my simple webserver needs to serve files from the filesystem and so I have needed to provide binary handlers, but the biggest problem I faced in doing this was that your compile-paths function currently forbids multi-field paths in the route.

SO I have made the smallest possible changes to the path->keys and compile-path functions to allow the use of Rails style multi-field params in the request URLs.

I have added a number of modules to provide tests of the old/new behaviour of both functions to ensure that I have made the correct changes.

I have also added a short summary of the new multi-field params in the Params section of your README.md to alert users to it possible use.

Let me know if there are any corrections you might want to be made to this pull request.

Regards,
Stephen Gaito

stephengaito committed:
failing first test ;-)

stephengaito committed:
added test-syntax-utils to allow for cleaner testing of private functions in external modules; added first tests of current behaviour for path->keys and compile-path

stephengaito committed:
fair tests of the expected behaviour for the changes to path->keys and compile-path functions

stephengaito committed:
updated README.md to add multi-fiedld wildcard parameters

stephengaito committed:
updated multiple multi-field wildcard documentation

dmac commented on README.md:
````diff
@@ -52,6 +53,41 @@ $ curl "http://localhost:8000/hi/Peter"
 Hello, Peter!
 ```
 
+OR using a multi-field "wildcard":
````
s/OR/or/

dmac commented on README.md:
````diff
@@ -52,6 +53,41 @@ $ curl "http://localhost:8000/hi/Peter"
+```
+
+OR any mixture of either single ('/'-delimited) fields or multi-field
+wildcards.
````
I think the additional examples from this line down are unnecessary; it shouldn't be surprising that both single-field and multi-field wildcards can work in the same url.

dmac commented on tests/path2keys.rkt:
```diff
@@ -0,0 +1,110 @@
+#lang racket
```
These are good tests, thanks for adding them! Can you rename this file to be main_test.rkt, since it's testing functionality defined in main.rkt?

dmac commented on tests/path2keys.rkt:
```diff
@@ -0,0 +1,110 @@
+#lang racket
+
+;; A simple collection of tests to verify the correct behaviour of the 
+;; complementary path->keys and compile-path functions
```
I don't think this comment, or any of the "verify old/new behavior" comments, are necessary.

dmac commented on tests/path2keys.rkt:
```diff
@@ -0,0 +1,110 @@
+;; complementary path->keys and compile-path functions
+
+(require rackunit)
+(require (for-syntax "./test-syntax-utils.rkt"))
```
Would RackUnit's `require/expose` remove the need for this readtable munging?

dmac commented on tests/path2keys.rkt:
```diff
@@ -0,0 +1,110 @@
+  '((id1 . "anId1") (id2 . "anId2")) )
+(check-equal? (local-request->key-bindings
+  "/aPathBase/:id1/aPathPart/:id2" "/aPathBase/anId1/aPathPart/anId2")
+  '((id1 . "anId1") (id2 . "anId2")) )
```
Extra space before the last parenthesis (many places).

dmac commented on tests/path2keys.rkt:
```diff
@@ -0,0 +1,110 @@
+    (id2 . "anId2")
+    (subPath2 . "aPathPart3/aPathPart4")
+  )
+)
```
Pull these close parentheses onto the last line with content.

dmac commented on tests/path2keys.rkt:
```diff
@@ -0,0 +1,110 @@
+;;
+;; Instead of using a request object we simply use the compiled-path and
+;; url.
+;;
```
Modifying the required code sounds complex and unnecessary. Can you help me understand the purpose of doing so?

stephengaito replied:
Daniel,

Many thanks for your careful comments.

I have implemented (as far as I can tell) all of your comments.

As you can probably tell, Racket/Lisp is not my "native" language.

It does however represent an important language for me for deep 
theoretical reasons.

As you can see I have heavily refactored what was essentially a test of 
the request->key-bindings function. Now instead of locally redefining 
it, I have added a few helper functions to make direct testing of the 
request->key-bindings function possible.  This provides a higher 
confidence that my changes do not break your code and indeed enhance it 
correctly.  Many thanks for pushing me to do this.

Regards,

Stephen Gaito

On 01/06/16 01:18, Daniel MacDougall wrote:
>
> In tests/path2keys.rkt 
> <https://github.com/dmac/spin/pull/8#discussion_r65285151>:
>
> > +  "^/aPathBase/([^/?]+)/([^?]+)(?:$|\\?)" )
> > +(check-equal? (compile-path "/aPathBase/*subPath/:id")
> > +  "^/aPathBase/([^?]+)/([^/?]+)(?:$|\\?)" )
> > +(check-equal? (compile-path "/aPathBase/:id1/*subPath/:id2")
> > +  "^/aPathBase/([^/?]+)/([^?]+)/([^/?]+)(?:$|\\?)" )
> > +(check-equal? (compile-path "/aPathBase/:id1/*subPath1/:id2/*subPath2")
> > +  "^/aPathBase/([^/?]+)/([^?]+)/([^/?]+)/([^?]+)(?:$|\\?)" )
> > +
> > +
> > +;; We now want to verify old behaviour on the construction of the
> > +;; url-pairs. To keep things as simple as possible, we make small changes
> > +;; to a copy of the relevant code from ../main.rkt and test it locally.
> > +;;
> > +;; Instead of using a request object we simply use the compiled-path and
> > +;; url.
> > +;;
>
> Modifying the required code sounds complex and unnecessary. Can you 
> help me understand the purpose of doing so?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/dmac/spin/pull/8/files/7bb206719b31c67d99a44710f77fa530f0c6960c#r65285151>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/ABBztGqfkn-IxgE03g7Wp4EPg5TXsHDMks5qHM_zgaJpZM4IkvPc>.
>

dmac replied:
I don't see your latest round of changes, are you sure you pushed your updates to this branch?

stephengaito replied:
Daniel,

Indeed I just realized that I actually applied these changes to the 
wrong pull request ;-(

Can you tell that I do not do enough multi-branch work nor pull requests :$

So... I will learn enough git to figure out how to rebase my last commit 
to the correct branch and then send you an email to let you know what I 
am done.  (This *is* something git has been designed to do... so it 
should not be difficult.... just an additional learning experience for me).

Regards,

Stephen Gaito

On 03/06/16 18:46, Daniel MacDougall wrote:
>
> In tests/path2keys.rkt 
> <https://github.com/dmac/spin/pull/8#discussion_r65747860>:
>
> > +  "^/aPathBase/([^/?]+)/([^?]+)(?:$|\\?)" )
> > +(check-equal? (compile-path "/aPathBase/*subPath/:id")
> > +  "^/aPathBase/([^?]+)/([^/?]+)(?:$|\\?)" )
> > +(check-equal? (compile-path "/aPathBase/:id1/*subPath/:id2")
> > +  "^/aPathBase/([^/?]+)/([^?]+)/([^/?]+)(?:$|\\?)" )
> > +(check-equal? (compile-path "/aPathBase/:id1/*subPath1/:id2/*subPath2")
> > +  "^/aPathBase/([^/?]+)/([^?]+)/([^/?]+)/([^?]+)(?:$|\\?)" )
> > +
> > +
> > +;; We now want to verify old behaviour on the construction of the
> > +;; url-pairs. To keep things as simple as possible, we make small changes
> > +;; to a copy of the relevant code from ../main.rkt and test it locally.
> > +;;
> > +;; Instead of using a request object we simply use the compiled-path and
> > +;; url.
> > +;;
>
> I don't see your latest round of changes, are you sure you pushed your 
> updates to this branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/dmac/spin/pull/8/files/7bb206719b31c67d99a44710f77fa530f0c6960c#r65747860>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/ABBztJ_KIY6nImCZwsXFJDIUPafox5fPks5qIGiKgaJpZM4IkvPc>.
>

dmac commented:
Hi @stephengaito, thanks for this feature! I think it will be a nice addition after we address a few things.

dmac commented on README.md:
````diff
@@ -52,6 +53,41 @@ $ curl "http://localhost:8000/hi/Peter"
+```
+
+```
+$ curl "http://localhost:8000/appFiles/images/jasmine_favicon.png"
````
s/jasmine_favicon.png/image.png/

dmac commented on tests/path2keys.rkt:
```diff
@@ -0,0 +1,110 @@
+(include "../main.rkt")
+
+;; verify the old behaviour
+(check-equal? (path->keys "/aPathBase/id") '())
```
s/aPathBase/base/

dmac replied:
(and similar simplifications throughout)

stephengaito committed:
completed changes as suggested by Daniel MacDougall; heavily refactored
the testing of the request->key-bindings function to make use of existing
infrastructure as supplied by main.rkt

(cherry picked from commit 4b956f102a8c97e7fe315411528cb17f7551c5e2)
(This corrects my performing the above changes on the wrong pull request.
 The changes were applied to the adding-binary-servlets branch/pull-request
 BUT should have been applied to the adding-multi-segment-route-parameters
 branch/pull-request)
